google/resumable_media/_upload.py

```diff
@@ -860,6 +869,375 @@ def recover(self, transport):
+            encrypted data. These headers will be propagated to individual
+            XMLMPUPart objects spawned from this container as well.
+
+    Attributes:
+        upload_url (str): The URL where the content will be uploaded.
+    """
+
+    def __init__(self, upload_url, filename, headers=None, upload_id=None, parts=None):
+        super().__init__(upload_url, headers=headers)
+        self._filename = filename
+        self._content_type = None
+        self._upload_id = upload_id
+        self._parts = parts if parts is not None else {}
+
+    @property
+    def upload_id(self):
+        return self._upload_id
+
+    def register_part(self, part_number, etag):
+        self._parts[part_number] = etag
+
+    def _prepare_initiate_request(
+        self, content_type
+    ):  # FIXME: figure out metadata in headers or add custom headers
+        """Prepare the contents of HTTP request to initiate upload.
+
+        This is everything that must be done before a request that doesn't
+        require network I/O (or other I/O). This is based on the `sans-I/O`_
+        philosophy.
+
+        Args:
+            content_type (str): The content type of the resource, e.g. a JPEG
+                image has content type ``image/jpeg``.
+
+        Returns:
+            Tuple[str, str, bytes, Mapping[str, str]]: The quadruple
+
+              * HTTP verb for the request (always POST)
+              * the URL for the request
+              * the body of the request
+              * headers for the request
+
+        Raises:
+            ValueError: If the current upload has already been initiated.
+
+        .. _sans-I/O: https://sans-io.readthedocs.io/
+        """
+        if self.upload_id is not None:
+            raise ValueError("This upload has already been initiated.")
+
+        self._content_type = content_type
+
+        initiate_url = (
+            self.upload_url + _MPU_INITIATE_QUERY
+        )  # FIXME: consider switch to urlparse
+
+        return _POST, initiate_url, None, self._headers
+
+    def _process_initiate_response(self, response):
+        """Process the response from an HTTP request that initiated upload.
+
+        This is everything that must be done after a request that doesn't
+        require network I/O (or other I/O). This is based on the `sans-I/O`_
+        philosophy.
+
+        This method takes the URL from the ``Location`` header and stores it
+        for future use. Within that URL, we assume the ``upload_id`` query
+        parameter has been included, but we do not check.
+
+        Args:
+            response (object): The HTTP response object (need headers).
+
+        .. _sans-I/O: https://sans-io.readthedocs.io/
+        """
+        _helpers.require_status_code(response, (http.client.OK,), self._get_status_code)
+        root = ElementTree.fromstring(response.text)
+        self._upload_id = root.find(_S3_COMPAT_XML_NAMESPACE + _UPLOAD_ID_NODE).text
```

Comment:
I think that's fine because if the server returns invalid XML according to its own definition of the endpoint output, then we can't recover from that and will need to raise an exception, which this will do.